### ISSUE TITLE
Broken pipe when running on a CI 

### ISSUE THREAD

**Issue by Enders** [thoughtbot/capybara-webkit#719]
Hello,

I have been banging my head against this for a few days and would love some guidance on how I can debug this. I am running:

- Capybara-webkit 1.4.1
- Capybara 2.4.4
- Ruby 2.2.0
- Rspec 3.1.0

When I run the test (single one) locally, it works, when I run the same thing on the CI through a SSH (not even through the TeamCity CI we are using) I get a Broken Pipe. I'm using the Headless gem to get it to work. The whole setup was working on a different CI, I just moved to new hardware and reinstalled from scratch.

Here is the output with `:webkit_debug`:

```
Finished "EnableLogging()" with response "Success()" 
Wrote response true "" 
Received "Visit(http://127.0.0.1:9887/)" 
Started "Visit(http://127.0.0.1:9887/)" 
Load started 
"Visit(http://127.0.0.1:9887/)" started page load 
Started request to "http://127.0.0.1:9887/" 
Finished "Visit(http://127.0.0.1:9887/)" with response "Success()" 
Started request to "http://127.0.0.1:9887/assets/signed-out.css" 
Started request to "http://127.0.0.1:9887/assets/signed_out.js" 
Received 200 from "http://127.0.0.1:9887/" 
Received 200 from "http://127.0.0.1:9887/assets/signed-out.css" 
Received 200 from "http://127.0.0.1:9887/assets/signed_out.js" 
Request to unknown URL: http://assets.zendesk.com/embeddable_framework/main.js
To block requests to unknown URLs:
  page.driver.block_unknown_urls
To allow just this URL:
  page.driver.allow_url("http://assets.zendesk.com/embeddable_framework/main.js")
To allow requests to URLs from this host:
  page.driver.allow_url("assets.zendesk.com")
Started request to "http://assets.zendesk.com/embeddable_framework/main.js" 
Started request to "http://127.0.0.1:9887/assets/logo-main-dark.png" 
Started request to "http://127.0.0.1:9887/assets/logos/logo_viadeo.png" 
Started request to "http://127.0.0.1:9887/assets/homepage/bg_feature1.jpg" 
Started request to "http://127.0.0.1:9887/assets/homepage/bg_feature2.jpg" 
Started request to "http://127.0.0.1:9887/assets/homepage/bg_feature3.jpg" 
Started request to "http://127.0.0.1:9887/assets/homepage/feature1_computer1.png" 
Started request to "http://127.0.0.1:9887/assets/homepage/feature1_computer2.png" 
Started request to "http://127.0.0.1:9887/assets/homepage/winddle-path2.png" 
Started request to "http://127.0.0.1:9887/assets/homepage/video-small.png" 

  An error occurred in an after hook
    Errno::EPIPE: Broken pipe
    occurred at /home/teamcity/.rvm/gems/ruby-2.2.0@winddle-rails4.2/gems/capybara-webkit-1.4.1/lib/capybara/webkit/connection.rb:30:in `write'
```

I have tried running it with Qt4 and Qt5, both 64bits and 32bits (that's a lot of apt-get remove / gem uninstall) and the error is always the same.

`dmesg` shows the following error, confirming it's an internal error:

```
[655552.072558] webkit_server[11965]: segfault at 0 ip           (null) sp 00007fffb6049178 error 14 in webkit_server[400000+54000]
```

That's with Qt5 64bits. With Qt4 the error was different:

```
[622593.029889] traps: webkit_server[20399] general protection ip:7f2cecf6d787 sp:7fff2564caf0 error:0 in libQtWebKit.so.4.10.2[7f2ceba8e000+1fc9000]
```

I understand that you cannot do much about QT internal bugs, but because it occurs on both QT4 and QT5 on a fresh install and with what I assume is a pretty standard test (it crashes right after starting to fetch the homepage's resources), there may be more to this than just 'random configuration issue that would never happen to anyone else', or so I keep hoping :)

Thank you!

**Comment by mhoran** [comment]
First off, thanks for taking all these steps to debug. I understand the difficulty and frustration involved. This looks like a genuine crash to me, especially under Qt 5. However, the Qt 4 failure could be masked by something else.

I've resurrected my `breakpad` branch, which can be used to upload crash reports. You should be able to get it by adding the following to your `Gemfile`:

```ruby
gem 'capybara-webkit', github: 'mhoran/capybara-webkit', branch: 'breakpad', submodules: true
```

When the crash occurs, it will print out a message to run the upload script. The crash dump should be located in `/tmp` but the printed error should tell you where it is. When you run the script, it will send a crash dump to me via email.

**Comment by Enders** [comment]
Hey there Matthew, thanks for taking the time to look at this :) 

I ran the script, hopefully you got the info you need? FYI, the script also gave this output:

```
/home/teamcity/.rvm/gems/ruby-2.2.0@winddle-rails4.2/bundler/gems/capybara-webkit-4a4370f9a850/bin/webkit_server, section '.eh_frame': the call frame entry at offset 0x18 sets the rule for register '.ra' to 'undefined', but the Breakpad symbol file format cannot express this
/home/teamcity/.rvm/gems/ruby-2.2.0@winddle-rails4.2/bundler/gems/capybara-webkit-4a4370f9a850/bin/webkit_server, section '.eh_frame': the call frame entry at offset 0x48 uses a DWARF expression to describe how to recover register '.cfa',  but this translator cannot yet translate DWARF expressions to Breakpad postfix expressions
```

Seemed to be a breakpad warning and that the file still got uploaded (no error messages related to the post to /crash_report).

**Comment by mhoran** [comment]
Awesome; it worked! I haven't tried running that code in about two years, so it's a bit of a surprise. I'm mostly impressed that the Heroku upload app still works.

Anyway, I didn't see any suspect capybara-webkit code in the stack. Every thread was somewhere in Qt land. Of course, that doesn't mean it's not our fault. However, I did notice a mix of Qt 5.2 and 5.1 libraries in the trace:

```
libQt5WebKit.so.5.1.1
libQt5Core.so.5.2.1
libQt5Widgets.so.5.2.1
```

Thread 0 crashed with a null pointer exception. I can't tell if the crash was in `webkit_server` or somewhere in Qt land. It looks like you don't have debugging symbols installed (or perhaps breakpad can't find them.) Here's the interesting bits of the trace:

```
Thread 0 (crashed)
 0  0x0
    rbx = 0x0000000000991e30   r12 = 0x00007fda0806f260
    r13 = 0xffff000000000002   r14 = 0x00007fda1bef5ff0
    r15 = 0x0000000000000000   rip = 0x0000000000000000
    rsp = 0x00007fff2c7310c8   rbp = 0x00007fda1bfbd628
    Found by: given as instruction pointer in context
 1  libQt5WebKit.so.5.1.1 + 0x683e6b
    rip = 0x00007fda1a020e6c   rsp = 0x00007fff2c7310d0
    Found by: stack scanning
```

I'd recommend first checking if you can get a uniform installation of Qt on the system. If that doesn't fix it, perhaps we can get debugging symbols installed and see if breakpad will pick them up. If it doesn't, then we may have to break out `gdb` on the system itself, which should pull in the debugging symbols. You'll want to keep running the breakpad branch, as it will compile the `webkit_server` with debugging symbols, which are required for `gdb`.

**Comment by Enders** [comment]
Hmm... I just spent the last day and a half working on this and I don't have much to show for it. Turns out that Ubuntu doesn't provide any coherent QT packages, so I tried to compile QT 5.4.0 from source and that was a nightmare due to the VPS being pretty bare. Everything but libWebkit compiled, due to some unicode issues (apparently, sizeof(UChar) != 2, cool story).

So, I downloaded the already compiled binaries for amd64 archi and after a few more apt-get install, I was able to get capybara-webkit running on a all 5.4.0 set of libraries instead of a mix of 5.2.1 / 5.1.1. The good news is that now **it mostly works!**

I haven't been able to get a green build yet, there is always a few dozens of Broken Pipe in the mix (only the js: true tests are failing) and re-running those tests show a green, so the issue is different from before. I can't say I'm super thrilled, but it's progress? 

I have uploaded the latest crash, hopefully, it contains something useful... Below is the output just before the info on where the dump is:

```
ASSERT failure in QList<T>::operator[]: "index out of range", file /usr/local/Qt-5.4.0/include/QtCore/qlist.h, line 482
```

Thanks!

**Comment by mhoran** [comment]
It definitely looks like something is going on inside of Qt itself, but it's hard to tell without a full stack trace. I haven't seen this one myself, though I've not run the versions that you're running. Our CI is on 4.8 and my development machine is on 5.3.

If you can get some additional tools installed on the CI box, we can try and get a backtrace from `gdb`. You'll need to get `gdb` installed from the package manager. Then, run `ulimit -c unlimited`. This will enable core dumps for the current shell. Once enabled, run the test suite and wait for a crash. When the process crashes, you should get a `core` file in the working directory. You can then use `gdb` to print a backtrace.

You'll need to know the path to `webkit_server` to process the core dump. You'll also need debugging symbols in the server, so be sure to run the `breakpad` branch. You can find the gem installation directory with bundle show. The `webkit_sever` binary will be in `bin`. Once you find the full path to the `webkit_server` binary, run `gdb <path to webkit_server> <path to core>`. This will dump you in a REPL. From here, type `bt`. This should return a useful trace from the crashed thread.

**Comment by Enders** [comment]
Matthew, thank you very much for this thorough follow-up!

As of now, builds are no longer failing reliably, which is great as far as this issue is concerned (Yay). Haven't had a green build yet (1-2 failures out of 5000+ tests), but the errors are now more in line with what I'm used to, even on the previous CI server and they are bounded to a single test so wekbit_server no longer crash.

The main thing that resolved the situation was your comment on the different versions of QT (5.1.1 for libqt5webkit5 / 5.2.1 for the rest), the actual solution for me was to **install QT 5.4.0 directly from QT's website (compiled binaries) rather than using Ubuntu Server 64bit 14.04 LTS packages**, which are flawed. 

I'll be closing this issue, and hopefully, the information put here will be good reference next time there is a crash to investigate :) Thanks again!

**Comment by bpinto** [comment]
@mhoran After running the tests using your branch I have the .dmp file in hand, how can I read it?

I can read some things, but most of it is like this `ؽP�nux-0�ؽ`9ؽ��nux-p�gؽpؽЧnux-9ؽ0�`

**Comment by mhoran** [comment]
Hey @bpinto. The dump is intended to be uploaded via the `upload_dump.rb` script and analyzed on our end. The `upload_dump.rb` script does some additional processing, but some other tools are required to extract a stack trace. Feel free to run the script (instructions should be printed as part of the crash.) I'll then be able to extract a stack trace.

**Comment by bpinto** [comment]
@mhoran Isn't there another way? I'm afraid I can't upload that dump to you as it contains environment variables that we can not share.

I am facing intermittent EOF erros followed by Broken Pipe errors on remaining capybara-webkit tests and I don't know what is causing them.
I saw some font related things on the dump and tried to do the "font fix" but it didn't solve, as I'm running the tests on a dockerized centos machine, I didn't expect it to be the solution.

**Comment by mhoran** [comment]
Instead of analzying the breakpad dump, you'll want to capture a core dump by setting `ulimit -c unlimited` and then using `gdb`. When you get a core dump, `gdb $WEBKIT_SERVER_PATH $CORE_DUMP_PATH ` will start a gdb session. Then `bt` should print out a useful trace of the crashed thread.

You'll want to do this from the breakpad branch, as it has debugging symbols enabled for `webkit_server`. However, it's quite possible that the crash is inside Qt (as was true for the original issue), in which case there's not much we can do. You may also want to look into upgrading your Qt version, if you're still running 4.8.

**Comment by bpinto** [comment]
I am running latest 5.40 qt installed from stock centos 7 yum –following wiki instructions.

I have a feeling it could be a javascript issue though as this error only happens to a single javascript enabled test suite.

I have tried doing that but it says that there is another process watching the trace or something like that. I believe it's because the breakpad branch is also "listening", is there a simple way to disable this from happening on the breakpad branch so that I can benefit from its debugging symbols?

**Comment by bpinto** [comment]
Okay, I was able to run it, these are the instructions if someone else opens this topic:
```
gdb -q `which ruby`
r `which bundle` exec rspec
```

Backtrace: http://dpaste.com/1WF5GZR

@mhoran Can you see if you find anything useful is this backtrace? Thanks!

**Comment by mhoran** [comment]
Hey @bpinto. If you modify the `webkit_server.pro` file in the root of the project to include `CONFIG += debug`, you should get debugging symbols. From the master branch, I would expect no issue regarding "another process watching the trace".

Happy to push another branch if you haven't already forked the project so you can add this on your own. However, simply forking and adding `debug` to the `CONFIG` line should do. Then `ulimit -c unlimited` and `gdb` as previously described should produce a useful backtrace.

Good to hear that you're running Qt 5.4. I've not tried that myself, but Qt tends to get more stable, not less, with newer releases.

**Comment by Masa331** [comment]
I just spent a lot of time on basically same error. I'v been able to reproduce the error on very simple app here: 
https://github.com/Masa331/webkit_test

It would fail every test run on broken pipe or connection reset by peer. However removing jquery or modernizr from app js manifest would fix it.

I tried both 64bit QT 4 and 5 from Ubuntu(i have Lubuntu 14.04 btw) packages and there wasn't any change. `dmesg` showed same errors as for @Enders .

**I can confirm that manually compiling latest QT(5.4.1) from it's website fixed the errors and i haven't seen them since.**
